lib/scholar/cluster/affinity_propagation.ex

```diff
@@ -0,0 +1,382 @@
+defmodule Scholar.Cluster.AffinityPropagation do
+  @moduledoc """
+  Model representing affinity propagation clustering.
+  """
+
+  import Nx.Defn
+  import Scholar.Shared
+
+  @derive {Nx.Container,
+           containers: [
+             :labels,
+             :cluster_centers_indices,
+             :affinity_matrix,
+             :cluster_centers,
+             :num_clusters,
+             :similarity_matrix
+           ]}
+  defstruct [
+    :labels,
+    :cluster_centers_indices,
+    :affinity_matrix,
+    :cluster_centers,
+    :num_clusters,
+    :similarity_matrix
+  ]
+
+  @opts_schema [
+    iterations: [
+      type: :pos_integer,
+      default: 300,
+      doc: "Number of iterations of the algorithm."
+    ],
+    damping_factor: [
+      type: :float,
+      default: 0.5,
+      doc: """
+      Damping factor in the range [0.5, 1.0) is the extent to which the
+      current value is maintained relative to incoming values (weighted 1 - damping).
+      """
+    ],
+    self_preference: [
+      type: {:or, [:float, :boolean, :integer]},
+      doc: "Self preference."
+    ],
+    seed: [
+      type: :integer,
+      doc: """
+      Determines random number generation for centroid initialization.
+      If the seed is not provided, it is set to `System.system_time()`.
+      """
+    ]
```

Comment:
I think it's better if we receive a `key` tensor instead.
This allows the algorithm to play nicer with other functions that also use Nx.Random

Reply:
The algorithm should also return the key -- or make it clear that the key should be manually split beforehand

Reply:
Let's keep seed for now, as this is a larger change, and we figure out a way of injecting either key or seed later?